net: ipv6: resolve literal IPv6 addresses

hubertmis opened the pull request:
With this patch the resolver module can resolve literal IPv6
addresses even when DNS client is not presnet in the system.

cfriedt commented:
It's looking good. Are you able to do include some additional tests?

jukkar commented on subsys/net/lib/sockets/getaddrinfo.c:
```diff
@@ -19,6 +21,22 @@ LOG_MODULE_REGISTER(net_sock_addr, CONFIG_NET_SOCKETS_LOG_LEVEL);
+ */
+#define INIT_ADDRINFO(addrinfo, sockaddr) { \
+		(addrinfo)->ai_addr = &(addrinfo)->_ai_addr; \
+		(addrinfo)->ai_addrlen = sizeof(*sockaddr); \
```
I know this is not your changes, but just to be in the safe side, the `()` is needed around sockaddr too, so `(addrinfo)->ai_addrlen = sizeof(*(sockaddr));`

jukkar commented on subsys/net/lib/sockets/getaddrinfo.c:
```diff
@@ -346,6 +353,86 @@ static inline int z_vrfy_z_zsock_getaddrinfo_internal(const char *host,
 
+#if defined(CONFIG_NET_IPV6)
+static int try_resolve_literal_ipv6(const char *host, const char *service,
+		const struct zsock_addrinfo *hints,
+		struct zsock_addrinfo *res)
```
Indentation is off, please align with the host parameter

jukkar commented on subsys/net/lib/sockets/getaddrinfo.c:
```diff
@@ -346,6 +353,86 @@ static inline int z_vrfy_z_zsock_getaddrinfo_internal(const char *host,
+		return DNS_EAI_NONAME;
+	}
+
+	if (host[0] != '[' || host[strlen(host) - 1] != ']') {
```
There already exits `net_ipaddr_parse()` that provides similar functionality (parsing IPv[4|6] string). Could we utilize that functionality here so we do not need to call any malloc etc.

hubertmis replied:
I have not found this function. Thanks for suggestion!

hubertmis replied:
It looks this module could be refactored and literal IPv4 addresses could be checked alongside IPv6, instead of within DNS procedure. `net_ipaddr_parse()` parses IPv6 and IPv4 at once.

I would leave it as `TODO`, because I don't have any applications to verify if literal IPv4 and DNS resolvers work as expected. Someone working with IPv4 could do that later.

hubertmis commented:
> Are you able to do include some additional tests?

I've created some tests based on RFC 2732

jukkar reviewed (approved):
LGTM

rlubos reviewed (commented):
Not sure about this one. Technically looks ok, but it adds some extra complexity to the logic and duplicates some code from `z_impl_z_zsock_getaddrinfo_internal` which could easily be avoided if that function was updated instead (line 208). I wouldn't be concerned about breaking literal IPv4 handling - there is an unit test that covers it.

jukkar commented on subsys/net/lib/sockets/getaddrinfo.c:
```diff
@@ -346,6 +354,65 @@ static inline int z_vrfy_z_zsock_getaddrinfo_internal(const char *host,
+
+	result = net_ipaddr_parse(host, strlen(host), &res->_ai_addr);
+
+	if (result && (net_sin6(&res->_ai_addr)->sin6_family == AF_INET6)) {
```
> You don't know at this point whether to use IPv4/IPv6 structure, it would be more appropriate to check against generic `sockaddr`, i. e. `res->_ai_addr.sa_family`

The sa_family fields in different structs point to same location (first two bytes of those structs), so it does not matter which family field we are using.

rlubos replied:
Yes, I understand that this won't make a difference to the compiler, I was more thinking of folks analyzing this code in the future - one may wonder why is IPv6 assumed before it's actually verified that we deal with IPv6.

jukkar replied:
Sure, it makes it more clear to the user if the generic sockaddr struct family value is used

jukkar commented:
> I wouldn't be concerned about breaking literal IPv4 handling - there is an unit test that covers it.

Ah, I missed that one. Then we could just implement this is a generic fashion without any special IPv6 function.

hubertmis commented:
@rlubos , Please have a look again. I've added one more commit that moves literal IPv4 address resolving out of `z_impl_z_zsock_getaddrinfo_internal()` to the new function created in this PR.

It looks that most of duplicates are removed now. The only one remaining is `service` resolving.

With this PR one can resolve literal IPv4 and IPv6 addresses when there is no DNS in the system/network.

jukkar commented:
Could you change the patches so that the IPv4/6 functionality is in one commit and the tests are in different commit.

hubertmis commented:
@jukkar, I'll squash IPv6 and IPv4 changes to a single commit when we all agree that we would like to introduce changes for both protocols. While doing so, I can move new unit tests to a separated commit. But before I perform this action, I would like to know @rlubos 's opinion if proposed changes are OK. I think Robert suggested something else.

rlubos reviewed (commented):
Well I was more thinking of including IPv6 literal parsing into `z_impl_z_zsock_getaddrinfo_internal` but I see that would need some more sophisticated ifdef slicing within the function to get this feature available w/o DNS_RESOLVER enabled.
I guess current approach is also acceptable, at least literal IP parsing is not scattered across the code now.

Commit:
net: resolve literal IP addresses even when DNS is disabled

With this patch the resolver module can resolve literal IPv6
and IPv4 addresses even when DNS client is not presnet in
the system.

Signed-off-by: Hubert Miś <hubert.mis@nordicsemi.no>

Commit:
tests: net: Test resolving literal IPv6 addresses

This patch adds a test for the network address resolver module
that verifies resolving literal IPv6 addresses.

Signed-off-by: Hubert Miś <hubert.mis@nordicsemi.no>

hubertmis commented:
> Could you change the patches so that the IPv4/6 functionality is in one commit and the tests are in different commit

Just rearranged. @jukkar, is it OK now?

> I guess current approach is also acceptable

Thanks @rlubos for the confirmation!
